src/agents/OIDCCredentialAgent.js

```diff
@@ -23,29 +26,57 @@ export default class OIDCCredentialAgent {
     return this._accessToken;
   }
 
-  getCredentials() {
+  async getCredentials() {
     if (this.credentialsPromise && this.credentialsExpire > new Date()) {
       return this.credentialsPromise;
     }
 
     // Call the oidcCredentials endpoint with the access token.
     const loginBaseUrl = 'https://login.taskcluster.net';
     const url = `${loginBaseUrl}/v1/oidc-credentials/${this.oidcProvider}`;
+    const mutex = new Mutex();
+
+    await mutex.lock(MUTEX_LOCK_KEY);
+
+    const session = localStorage.getItem(MUTEX_LOCK_KEY);
+
+    // If the session exists, it was set by the leader mutex. Grab the
+    // credentials from storage and pass them on.
+    if (session) {
+      const response = JSON.parse(session);
 
-    this.credentialsPromise = fetch(url, {
-      headers: {
-        Authorization: `Bearer ${this.accessToken}`
-      }
-    })
-    .then((response) => {
       this.credentialsExpire = new Date(response.expires);
+      this.credentialsPromise = Promise.resolve(response.credentials);
```

Comment:
You'll need to check the expiration here, right?  And go on to the `else` branch if expired..

Reply:
The expiration of current credentials is checked on L30. Without `credentialsPromise` defined, we are getting fresh credentials here from a new lock, and so we know the expires to be approximately 15 minutes in the future.

I can add an expiration check here if you like just for sanity. Please let me know if you think that would be nice to have.

Reply:
With the additional comments, I see the error of my reasoning :)